src/audio/mux.c

```diff
@@ -26,52 +26,366 @@
+#define printij(i, j, x) trace_mux(#x " = %u, i = %u, j = %u", x, i, j)
+
+// TODO: to be removed before release
+static void print_cfg_values(struct sof_mux_config *cfg)
+{
+	uint8_t i, j;
+
+	print(cfg->mode);
+	print(cfg->frame_format);
+	print(cfg->num_channels);
+	print(cfg->num_streams);
+
+	for (i = 0; i < cfg->num_streams; i++) {
+		printi(i, cfg->streams[i].num_channels);
+		printi(i, cfg->streams[i].pipeline_id);
+		for (j = 0; j < cfg->streams[i].num_channels; j++)
+			printij(i, j, cfg->streams[i].mask[j]);
+	}
+}
+
+static int mux_set_values(struct comp_data *cd, struct sof_mux_config *cfg)
+{
+	int i, j;
+
+	// TODO: to be removed before release
+	print_cfg_values(cfg);
 
-/* tracing */
-#define trace_mux(__e) trace_event(TRACE_CLASS_MUX, __e)
-#define trace_mux_error(__e)   trace_error(TRACE_CLASS_MUX, __e)
-#define tracev_mux(__e)        tracev_event(TRACE_CLASS_MUX, __e)
+	cd->mode = cfg->mode;
+	cd->num_channels = cfg->num_channels;
+	cd->frame_format = cfg->frame_format;
+
+	for (i = 0; i < cfg->num_streams; i++) {
+		cd->streams[i].num_channels = cfg->streams[i].num_channels;
+		cd->streams[i].pipeline_id = cfg->streams[i].pipeline_id;
+		for (j = 0; j < cfg->streams[i].num_channels; j++)
+			cd->streams[i].mask[j] = cfg->streams[i].mask[j];
+	}
+
+	return 0;
+}
 
 static struct comp_dev *mux_new(struct sof_ipc_comp *comp)
 {
+	struct comp_dev *dev;
+	struct sof_ipc_comp_mux *mux;
+	struct sof_ipc_comp_mux *ipc_mux = (struct sof_ipc_comp_mux *)comp;
+	struct comp_data *cd;
+
 	trace_mux("mux_new()");
 
-	return NULL;
+	if (IPC_IS_SIZE_INVALID(ipc_mux->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_MUX, ipc_mux->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		      COMP_SIZE(struct sof_ipc_comp_mux));
+	if (!dev)
+		return NULL;
+
+	mux = (struct sof_ipc_comp_mux *)&dev->comp;
+	memcpy(mux, ipc_mux, sizeof(struct sof_ipc_comp_mux));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
 }
 
 static void mux_free(struct comp_dev *dev)
 {
+	struct comp_data *cd = comp_get_drvdata(dev);
 
+	trace_mux("mux_free()");
+
+	rfree(cd);
+	rfree(dev);
 }
 
 /* set component audio stream parameters */
 static int mux_params(struct comp_dev *dev)
 {
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_mux("mux_params()");
+
+	dev->params.channels = cd->num_channels;
+	dev->params.frame_fmt = cd->frame_format;
 
 	return 0;
 }
 
+static int mux_ctrl_set_cmd(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_mux_config *cfg;
+	int ret = 0;
+
+	trace_mux("mux_ctrl_set_cmd(), cdata->cmd = 0x%08x", cdata->cmd);
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		cfg = (struct sof_mux_config *)cdata->data->data;
+
+		ret = mux_set_values(cd, cfg);
```

Comment:
@akloniex one more clarification. You are assuming that the mux config will be supplied as binary data through a command callback. Is this assumption valid? Dont you need the config when the new mux comp is created?

I think you should follow something similar to what the channel selector does where the config data is sent as binary data in the  IPC payload.